### ISSUE TITLE
Implement transferable/cloneable objects

### ISSUE THREAD

**Issue by jdm** [servo/servo#15021]
Spec: https://html.spec.whatwg.org/multipage/infrastructure.html#transferable-objects, https://html.spec.whatwg.org/multipage/infrastructure.html#cloneable-objects

We currently do not pass pointers to JSStructuredCloneCallbacks when calling [JS_WriteStructuredClone](https://dxr.mozilla.org/servo/rev/9afd96024fc9cb3ea822e8d549f008e03177949b/components/script/dom/bindings/structuredclone.rs#31-37) and [JS_ReadStructuredClone](https://dxr.mozilla.org/servo/rev/d09bf70d220262e2c199abf61b4b78b7137d6665/components/script/dom/bindings/structuredclone.rs#65-71). These [callbacks](https://dxr.mozilla.org/mozilla-central/rev/de67fccc4c64a49f261aea29141357b94c7b3b9c/dom/base/StructuredCloneHolder.cpp#56-143) control what to do when DOM objects are encountered in the structured clone algorithm. We'll need to duplicate what Gecko does to support this in Servo.

**Comment by jdm** [comment]
This support is necessary for the transferable arguments to `postMessage`, which allows sending complex DOM objects between the main thread and workers.

**Comment by fflorent** [comment]
Hi @jdm! What are the requirements to be able to do take this issue?

Florent

**Comment by jdm** [comment]
Ability to read Gecko's C++ code (as linked in the original comment) and translate that into equivalent Rust code, mainly!

**Comment by KiChjang** [comment]
This would be a nice thing to have for `MessagePort` as well. We currently have some code that deals with structured clones in `components/script/dom/bindings/structuredclone.rs`.

**Comment by gterzian** [comment]
Is someone already working on this one? If not, I would be happy to give it a try...

**Comment by jdm** [comment]
Let's give @fflorent the right of first refusal here.

**Comment by KiChjang** [comment]
I'd also highly encourage the person to work on this discuss the implementation plan here first before attempting.

**Comment by fflorent** [comment]
Sure take it! :)

Florent

**Comment by gterzian** [comment]
@fflorent thanks! 

@KiChjang @jdm I'll take some time to look into this, and will come back to discuss the implementation first :)

**Comment by gterzian** [comment]
Hi there, I've started looking into this, and the first question I have is: The stuff found in https://github.com/servo/rust-mozjs/, are those just stubs or actual implementations as well? For example I see Servo is using [JS_WriteStructuredClone](https://github.com/servo/rust-mozjs/blob/master/src/jsapi_macos_64.rs#L6126) but I can't find the actual implementations. 

Same thing for [JSStructuredCloneCallbacks](https://github.com/servo/rust-mozjs/blob/master/src/jsapi_macos_64.rs#L2246), are those actually implemented somewhere, or not?

Looking at the Firefox code, I can see the actual callbacks implementation, so I'm wondering, does this issue require implementing those callbacks? If so, would that be done in `rust_mozjs`? Or are those already implemented and do they just need to be glued into Servo?

**Comment by KiChjang** [comment]
@gterzian rust-mozjs are Rust bindings to the JavaScript engine that we use called SpiderMonkey. It is essentially the js directory in Gecko.

**Comment by gterzian** [comment]
@KiChjang thanks for the info. Why is it that those JSStructuredCloneCallbacks are found within `rust_mozjs`, while in gecko they are in the `dom` folder, not `js`?

**Comment by gterzian** [comment]
Ok so here is what I can make out of the Gecko code:

* There is this [`dom/base/StructuredCloneHolder`](https://dxr.mozilla.org/mozilla-central/source/dom/base/StructuredCloneHolder). 
* Sometimes directly, sometimes through a lot of indirection, this `StructuredCloneHolder` ends up calling the various [`JS_*` operations in js/public/StructuredClone.h](https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h)
* `StructuredCloneHolder` also defines a bunch of `StructuredCloneHolder::sCallbacks`, which are basically implementations of [`JSStructuredCloneCallbacks`](https://dxr.mozilla.org/mozilla-central/source/js/public/StructuredClone.h#176)
* When calling the various `JS_*` operations, it passes around those `StructuredCloneHolder::sCallbacks`
* When those callbacks end up being called, they in turn call `handler` operations back in `dom/base/StructuredCloneHolder`. For example `StructuredCloneCallbacksRead` will call CustomReadHandler, and return the value back to the caller of the callback...
* There are also a bunch of [`StructuredCloneTags`](https://dxr.mozilla.org/mozilla-central/source/dom/base/StructuredCloneTags.h), which are used in the handler operations to handle various types of cloned values being read/written. 

So basically, like Gecko, we need to:

* implement those callbacks as a struct , using the `JSStructuredCloneCallbacks` types found here https://github.com/servo/rust-mozjs/blob/master/src/jsapi_macos_64.rs#L2246
* Pass this struct around instead of the current `ptr::null()` when calling operations like `JS_WriteStructuredClone` and `JS_ReadStructuredClone`
* Implement appropriate handlers for each callbacks. 

So my question now is, am I getting it somewhat right? And also, am I right to think that we should probably just start with one or two handler operations? I believe you mentioned two potential use cases, so we should probably just do those for now?

**Comment by jdm** [comment]
Yep, that sounds like a correct summary of how all the pieces fit together. I think implementing stubs for all of the callbacks makes the most sense; we will presumably need some kind of delegation like https://dxr.mozilla.org/mozilla-central/rev/1d025ac534a6333a8170a59a95a8a3673d4028ee/dom/base/StructuredCloneHolder.cpp#62-65, and then we can focus on supporting a single small use case like using `postMessage` to clone a `Blob` object when communicating with a worker.

**Comment by gterzian** [comment]
@jdm awesome, thanks!

**Comment by gterzian** [comment]
first pass at https://github.com/servo/servo/pull/15519, if you could please take a look...
